source/user/mcscf.rst

```diff
@@ -1,57 +1,299 @@
+
+
+1) (Default) Specifying no additional information.
+  This is the most minimal strategy for selecting an active space and chooses orbitals (and electrons) around the fermi-level that match the number of orbitals and electrons specified by the user.
+  In most circumstances, this is not an ideal strategy and will lead to poor convergence or none at all.
+  For example:
+.. code-block:: python
+
+  ncas, nelecas = (6,8)
+  mycas = myhf.CASSCF(ncas, nelecas)
+
+
+2) Specifying the molecular orbital (MO) index of the active space orbitals you want. 
+  The user can "manually" selects the MO orbital indices (in a 1-based indexing scheme) and passes them to the ``sort_mo`` function.
+  See :source:`examples/mcscf/10-define_cas_space.py` for more details.
+
+.. code-block:: python
+
+  mycas = mcscf.CASSCF(myhf, 4, 4)
+  # Note sort_mo by default take the 1-based orbital indices.
+  mo = mycas.sort_mo([5,6,8,9])
+  mycas.kernel(mo)
+
+
+3) Specifying the number of orbitals in each symmetry group. 
+This strategy can occasionally be helpful when the initial guess orbitals are not easily identifiable.
+
+.. code-block:: python
+
+  mycas = mcscf.CASSCF(mf, 12, 12)
+  ncore = {'A1g':5, 'A1u':5}
+  ncas = {'A1g':2, 'A1u':2,'E1ux':1, 'E1uy':1, 'E1gx':1, 'E1gy':1,
+              'E2ux':1, 'E2uy':1, 'E2gx':1, 'E2gy':1}
+  mo = mcscf.sort_mo_by_irrep(mycas, mf.mo_coeff, ncas, ncore)
+  mycas.kernel(mo)
+
+A similar approach where we specify the electron occupations by irreducible representation is also possible by setting ``mycas.fcisolver.irrep_nelec``.
+
+.. code-block:: python
+
+  mycas = mcscf.CASSCF(myhf, 8, 8)
+  mycas.fcisolver.irrep_nelec = {"A1g": (2, 1), "A1u": (1, 1), "E1ux": (1, 1), "E1uy": (1, 0)}
+
+4) Use automated strategies (``avas`` and ``dmet_cas``) to pick an active space based on AO orbitals you're targeting.
+For more details, see :source:`examples/mcscf/43-avas.py` and :source:`examples/mcscf/43-dmet_cas.py`.
+
+.. code-block:: python
+
+  from pyscf.mcscf import avas
+  ao_labels = ['Fe 3d', 'Fe 4d', 'C 2pz']
+  ncas, nelecas, orbs = avas.avas(mf, ao_labels)
+  mycas = mcscf.CASSCF(mf, ncas, nelecas)
+
+  
+
+.. code-block:: python
+
+  from pyscf.mcscf import dmet_cas
+  ao_labels = ['Fe 3d', 'Fe 4d', 'C 2pz']
+  ncas, nelecas, mo = dmet_cas.guess_cas(mf, mf.make_rdm1(), ao_labels)
+  mycas = mcscf.CASSCF(mf, ncas, nelecas)
+  mycas.kernel(mo)
+
+
+Frozen Core MCSCF
+-----------------
+
+To reduce to computational expense of CASSCF calculations, users can "freeze" orbitals thereby excluding them from optimization.
+
+Users can specify a number to free the N lowest orbitals:
+
+.. code-block:: python
+
+  mycas = mcscf.CASSCF(myhf, 6, 8)
+  mycas.frozen = 2
+  mycas.kernel()
+
+
+Users can also specify a list of orbital indices (0-based).
+These may be occupied, virtual, or active orbitals.
```

Comment:
Active orbitals are by definition not frozen!

Reply:
While I agree that active orbitals should never really be frozen, I found that they can be when looking at the `examples/19-frozen_core.py`, see below:

https://github.com/pyscf/pyscf/blob/7be5e015b2b40181755c71d888449db936604660/examples/mcscf/19-frozen_core.py#L36-L40

If people think it will just confuse users, I can say that only occupied or virtual orbitals can be frozen even though that's not strictly true.

Reply:
No, the point is that that description is wrong. If you assign the orbitals to the frozen core, then they are frozen even though they would be otherwise chosen as active.

I think it's a good point to make that the code is flexible enough to do this: it might be that your system has some orbitals close to the Fermi level that aren't really entangled, so freezing them saves a lot of time.

Reply:
So the correct expression is that one can also freeze orbitals that are either occupied or virtual in the reference determinant that are close to the Fermi level.

Reply:
Frozen means not optimized during SCF. So I think active orbitals can be frozen, i.e. not optimized, yet their occupation will not be restricted in the wave-function expansion, i.e. they are still active.

Reply:
Exactly, `frozen` keyword controls restrictions of orbital optimization, but has nothing to do with orbital restriction, which is controlled by active space definition. The keyword is smart in PySCF, if you give a number, it will define a **frozen core**, but if you give a list, it will exclude from orbital optimization orbital from any space.

Reply:
@bogdanoff no, the exact opposite: https://sunqm.github.io/pyscf/ci.html
```
frozen
: int or list

    If integer is given, the inner-most orbitals are frozen from CI amplitudes. 
```

**Frozen core** means the orbitals do not participate in CI calculations. It is exactly used to define the active space. For instance, if you freeze the 1s orbitals in H2O, you're left with 8 electrons. For a planar system like a polyacene, one typically freezes the sigma electrons, and correlates only the pi electrons. This is the use case where you want to set the frozen orbitals as a list.

Reply:
This is an easy point of confusion, since different codes have different definitions. For instance, in Psi4 you can 1) freeze the core with the `freeze_core` setting, **or** 2) set the orbitals separately, where they split to a) FROZEN orbitals, b) restricted orbitals, c) active orbitals, d) inactive orbitals, e) frozen virtuals [you can also define the active and inactive orbitals with RAS spaces]. Frozen orbitals are either doubly occupied or not unoccupied, and they are not optimized. Restricted and inactive orbitals are also either doubly occupied or unoccupied; however, they *are* included in the orbital optimization.

So, the difference is between frozen core, and frozen orbital.

Reply:
This is also telling us that the difference between frozen core and frozen orbital needs to be discussed thoroughly in the documentation.

Feel free to copy-paste my explanation above.

Reply:
Aaaandd... I now see your point, @bogdanoff: this is indeed an argument for the optimizer :see_no_evil:

... which means that I am still right, only in the totally opposite sense :rofl:

Reply:
Indeed, one has to use of word *frozen* with care.
In Molpro, there is a `freeze, orb.sym` keyword in mcscf module, which prevents orbital `orb.sym` from optimization. But no word *frozen* is directly used for active orbitals.
Just for the record text from  [https://sunqm.github.io/pyscf/mcscf.html](https://sunqm.github.io/pyscf/mcscf.html):
```
frozen
: int or list
    If integer is given, the inner-most orbitals are excluded from optimization.
    Given the orbital indices (0-based) in a list, any doubly occupied core orbitals,
    active orbitals and external orbitals can be frozen.
```
Probably the most important use of freezing active orbitals is x-ray spectroscopy with hole in deep core orbital that otherwise will be optimized away from the active space.